src/rgw/rgw_rest_s3.cc

```diff
@@ -6152,6 +6152,11 @@ int RGWSelectObj_ObjStore_S3::run_s3select(const char* query, const char* input,
 
   if (m_result.size() > strlen(PAYLOAD_LINE)) {
     m_result.append(END_PAYLOAD_LINE);
+    // create_message actully uses 16 bytes for encoding(4) ,which is 4 bytes more than actual size of string; string-capacity is much bigger than string-size(doubled) for most cases
+    // but in some cases it might be the same(size == capacity).
+    if(m_result.capacity() - m_result.size()<4){
+      m_result.reserve(m_result.size()+4);
+    }
```

Comment:
i'm confused by the buffer management here. `create_message()` intends to append 4 ints (16 bytes) at the end of this buffer? where does the `m_result.size() - 12` come from that's being passed to `create_message()` and why does this only need to add 4 bytes?

regardless, we shouldn't write any bytes past the string's size, so this should call resize() instead of messing with capacity/reserve here:
```c++
m_result.resize(m_result.size()+4);
```

Reply:
> i'm confused by the buffer management here. `create_message()` intends to append 4 ints (16 bytes) at the end of this buffer? where does the `m_result.size() - 12` come from that's being passed to `create_message()` and why does this only need to add 4 bytes?
> 
> regardless, we shouldn't write any bytes past the string's size, so this should call resize() instead of messing with capacity/reserve here:
> 
> ```c++
> m_result.resize(m_result.size()+4);
> ```

the 3 integers (encode_int) are at the beginning of the buffer (according to spec,  otherwise it won't work)
the last(No 4) is at the end of the message(according to spec).

Reply:
okay, thanks!

if `create_message()` needs to extend this buffer, maybe it would be better to pass a reference to the std::string instead of the char* pointer so it can resize the buffer itself?

Reply:
The encode_int is binary (contain null)

Reply:
null characters are fine. std::string keeps track of its size